File: apps/forge/lib/api/governance.ts
/**
 * Governance Center — typed shapes + query keys (Step-72).
 *
 * Mirrors `backend/app/schemas/governance.py` (camelCase wire format
 * via Pydantic `serialize_by_alias=True`). The pair
 * `governance.ts` + `governance-hooks.ts` follows the `dashboard.ts`
 * + `dashboard-hooks.ts` convention.
 */

// ---------------------------------------------------------------------------
// Policies
// ---------------------------------------------------------------------------

export type PolicyStatus = 'active' | 'archived' | 'draft';
export type PolicySeverity = 'low' | 'medium' | 'high' | 'critical';
export type PolicyCategory = 'security' | 'compliance' | 'cost' | 'privacy';

export interface PolicyActor {
  readonly id: string;
  readonly displayName: string;
}

export interface PolicyRead {
  readonly id: string;
  readonly title: string;
  readonly summary: string;
  readonly status: PolicyStatus;
  readonly severity: PolicySeverity;
  readonly category: PolicyCategory;
  readonly version: string;
  readonly updatedAt: string;
  readonly updatedBy: PolicyActor;
}

// ---------------------------------------------------------------------------
// Approvals
// ---------------------------------------------------------------------------

export type ApprovalState = 'pending' | 'accepted' | 'declined' | 'expired';
export type ApprovalKind =
  | 'request_confirmation'
  | 'request_checkbox_confirmation'
  | 'ask_user_questions'
  | 'suggest_tasks';

export interface ApprovalRead {
  readonly id: string;
  readonly kind: ApprovalKind;
  readonly title: string;
  readonly prompt: string;
  readonly state: ApprovalState;
  readonly createdAt: string;
  readonly idempotencyKey: string;
  readonly decider: { id: string; displayName: string } | null;
  readonly decidedAt: string | null;
  readonly reason: string | null;
}

// ---------------------------------------------------------------------------
// RBAC roles
// ---------------------------------------------------------------------------

export interface RbacPermission {
  readonly resource: string;
  readonly actions: ReadonlyArray<string>;
}

export interface RbacRoleRead {
  readonly id: string;
  readonly name: string;
  readonly description: string | null;
  readonly permissions: ReadonlyArray<RbacPermission>;
  readonly memberCount: number;
  readonly system: boolean;
  readonly updatedAt: string;
}

// ---------------------------------------------------------------------------
// Board confirmations
// ---------------------------------------------------------------------------

export type BoardConfirmationOutcome = 'pending' | 'accepted' | 'declined';

export interface BoardConfirmationRead {
  readonly id: string;
  readonly subject: { id: string; identifier: string };
  readonly planRev: string;
  readonly outcome: BoardConfirmationOutcome;
  readonly decider: { id: string; displayName: string } | null;
  readonly decidedAt: string | null;
  readonly idempotencyKey: string;
  readonly prompt: string;
}

// ---------------------------------------------------------------------------
// Violations (LiteLLM-derived)
// ---------------------------------------------------------------------------

export type ViolationSeverity = 'low' | 'medium' | 'high' | 'critical';
export type ViolationStatus = 'OPEN' | 'RESOLVED' | 'REOPENED';

export interface Violation {
  readonly id: string;
  readonly timestamp: string | null;
  readonly model: string | null;
  readonly severity: ViolationSeverity;
  readonly kind: string;
  readonly description: string;
  readonly actor: string | null;
  readonly key_alias: string | null;
  readonly status: ViolationStatus;
}

export interface ViolationPollResult {
  readonly polled_at: string;
  readonly previous_poll_at: string | null;
  readonly count: number;
  readonly items: ReadonlyArray<Violation>;
}

// ---------------------------------------------------------------------------
// Audit
// ---------------------------------------------------------------------------

export interface AuditEvent {
  readonly id: string;
  readonly ts: string;
  readonly actor: string;
  readonly action: string;
  readonly target_type: string;
  readonly target_id: string;
  readonly tenant_id: string;
  readonly project_id: string;
  readonly metadata: Readonly<Record<string, unknown>>;
}

export interface AuditPage {
  readonly items: ReadonlyArray<AuditEvent>;
  readonly total: number;
  readonly page: number;
  readonly page_size: number;
}

// ---------------------------------------------------------------------------
// Query keys
// ---------------------------------------------------------------------------

export const governanceQueryKeys = {
  all: () => ['governance'] as const,
  policies: () => ['governance', 'policies'] as const,
  approvals: () => ['governance', 'approvals'] as const,
  rbacRoles: () => ['governance', 'rbac-roles'] as const,
  boardConfirmations: () => ['governance', 'board-confirmations'] as const,
  violations: () => ['governance', 'violations'] as const,
  audit: (page: number) => ['governance', 'audit', page] as const,
};

File: backend/app/schemas/governance.py
"""Governance Center schemas (Step-72).

CamelCase wire format — the orchestrator-stub surface (apps/forge/bin/
orchestrator-stub.py make_governance()) is the contract these models
match. Pydantic v2 + ``by_alias=True`` at the route produces camelCase
on the wire (R4 — typed artifacts, no free-form dicts).
"""

from __future__ import annotations

from datetime import datetime
from enum import StrEnum
from typing import Any
from uuid import UUID

from pydantic import ConfigDict, Field

from app.schemas.common import ForgeBaseModel


class _CamelModel(ForgeBaseModel):
    """Base that emits camelCase field names on serialization."""

    model_config = ConfigDict(
        from_attributes=True,
        populate_by_name=True,
        serialize_by_alias=True,
    )


# ---------------------------------------------------------------------------
# Policies
# ---------------------------------------------------------------------------


class PolicyStatus(StrEnum):
    ACTIVE = "active"
    ARCHIVED = "archived"
    DRAFT = "draft"


class PolicySeverity(StrEnum):
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"


class PolicyCategory(StrEnum):
    SECURITY = "security"
    COMPLIANCE = "compliance"
    COST = "cost"
    PRIVACY = "privacy"


class PolicyRead(_CamelModel):
    id: str
    title: str
    summary: str
    status: PolicyStatus
    severity: PolicySeverity
    category: PolicyCategory
    version: str
    updated_at: datetime = Field(..., alias="updatedAt")
    updated_by: dict[str, Any] = Field(..., alias="updatedBy")


class PolicyAcceptRequest(_CamelModel):
    actor_id: str | None = Field(default=None, alias="actorId")


# ---------------------------------------------------------------------------
# Approvals
# ---------------------------------------------------------------------------


class ApprovalState(StrEnum):
    PENDING = "pending"
    ACCEPTED = "accepted"
    DECLINED = "declined"
    EXPIRED = "expired"


class ApprovalKind(StrEnum):
    REQUEST_CONFIRMATION = "request_confirmation"
    REQUEST_CHECKBOX_CONFIRMATION = "request_checkbox_confirmation"
    ASK_USER_QUESTIONS = "ask_user_questions"
    SUGGEST_TASKS = "suggest_tasks"


class ApprovalRead(_CamelModel):
    id: str
    kind: ApprovalKind
    title: str
    prompt: str
    state: ApprovalState
    created_at: datetime = Field(..., alias="createdAt")
    idempotency_key: str = Field(..., alias="idempotencyKey")
    decider: dict[str, Any] | None = None
    decided_at: datetime | None = Field(default=None, alias="decidedAt")
    reason: str | None = None


class ApprovalDecisionRequest(_CamelModel):
    actor_id: str | None = Field(default=None, alias="actorId")
    reason: str | None = None


# ---------------------------------------------------------------------------
# RBAC roles
# ---------------------------------------------------------------------------


class RbacPermission(_CamelModel):
    resource: str
    actions: list[str]


class RbacRoleRead(_CamelModel):
    id: str
    name: str
    description: str | None = None
    permissions: list[RbacPermission]
    member_count: int = Field(default=0, alias="memberCount")
    system: bool = False
    updated_at: datetime = Field(..., alias="updatedAt")


# ---------------------------------------------------------------------------
# Board confirmations
# ---------------------------------------------------------------------------


class BoardConfirmationOutcome(StrEnum):
    PENDING = "pending"
    ACCEPTED = "accepted"
    DECLINED = "declined"


class BoardConfirmationRead(_CamelModel):
    id: UUID
    subject: dict[str, Any]
    plan_rev: str = Field(..., alias="planRev")
    outcome: BoardConfirmationOutcome
    decider: dict[str, Any] | None = None
    decided_at: datetime | None = Field(default=None, alias="decidedAt")
    idempotency_key: str = Field(..., alias="idempotencyKey")
    prompt: str


class BoardConfirmationAck(_CamelModel):
    """POST /board-confirmations body — board ack."""

    subject_id: str = Field(..., alias="subjectId")
    plan_rev: str = Field(..., alias="planRev")
    outcome: BoardConfirmationOutcome
    prompt: str | None = None
    idempotency_key: str | None = Field(default=None, alias="idempotencyKey")


__all__ = [
    "ApprovalDecisionRequest",
    "ApprovalKind",
    "ApprovalRead",
    "ApprovalState",
    "BoardConfirmationAck",
    "BoardConfirmationOutcome",
    "BoardConfirmationRead",
    "PolicyAcceptRequest",
    "PolicyCategory",
    "PolicyRead",
    "PolicySeverity",
    "PolicyStatus",
    "RbacPermission",
    "RbacRoleRead",
]
